File: lib/kobako/sandbox.rb
# frozen_string_literal: true

require "forwardable"

require_relative "errors"
require_relative "unresolved"
require_relative "sandbox_options"
require_relative "transport"
require_relative "catalog"
require_relative "context"

module Kobako
  # Kobako::Sandbox — the user-facing entry point for executing guest mruby
  # scripts inside a wasmtime-hosted Wasm module.
  #
  # The Sandbox owns the reusable configuration — the +Kobako::Runtime+ and
  # the +Kobako::Catalog::Services+ / +Catalog::Snippets+ / +Catalog::Extensions+
  # registries. Each +#eval+ / +#run+ seals that config on the first call and
  # drives one guest invocation through a fresh +Kobako::Context+ — its own
  # Handle table, dispatch +Proc+, captures, and usage — so the reusable
  # Sandbox holds no per-invocation state. The underlying wasmtime Engine and
  # compiled Module are cached at process scope by the native ext and never
  # surface to Ruby — constructing many Sandboxes amortises both costs
  # automatically.
  #
  # Output capture policy: the
  # per-channel cap (+stdout_limit+ / +stderr_limit+) is enforced inside the
  # WASI pipe — the host buffer stops growing at the cap, subsequent guest
  # writes on that channel fail or are dropped, and +#run+ still returns
  # normally. +#stdout+ / +#stderr+ return the captured prefix as a UTF-8
  # String; the byte content never carries a truncation sentinel.
  # +#stdout_truncated?+ / +#stderr_truncated?+ are the only way to observe
  # that the cap was hit.
  class Sandbox
    extend Forwardable

    attr_reader :wasm_path, :options

    # Per-option accessors forward to the immutable +SandboxOptions+ Value
    # Object so the Host App still reads them off Sandbox directly.
    def_delegators :@options, :timeout, :memory_limit, :stdout_limit, :stderr_limit, :profile

    # Returns the bytes the guest wrote to stdout during the most recent
    # invocation as a UTF-8 String, clipped at +stdout_limit+. Empty before
    # any invocation; the byte content never contains a truncation sentinel,
    # so use +#stdout_truncated?+ to observe overflow. Populated on every
    # outcome — including a rescued +TrapError+, after which it holds the
    # bytes written before the trap fired — mirroring +#usage+.
    def stdout = @last_context.stdout

    # Returns the bytes the guest wrote to stderr during the most recent
    # invocation as a UTF-8 String, clipped at +stderr_limit+. Empty before
    # any invocation. Mirror of +#stdout+.
    def stderr = @last_context.stderr

    # Returns +true+ iff stdout capture during the most recent invocation
    # exceeded +stdout_limit+. Resets to +false+ at the start of the next
    # invocation.
    def stdout_truncated? = @last_context.stdout_truncated?

    # Returns +true+ iff stderr capture during the most recent invocation
    # exceeded +stderr_limit+. Mirror of +#stdout_truncated?+.
    def stderr_truncated? = @last_context.stderr_truncated?

    # Returns the +Kobako::Usage+ value object for the most recent
    # invocation. Carries +wall_time+ (Float seconds the guest export call spent
    # inside wasmtime) and +memory_peak+ (Integer bytes, high-water of
    # the per-invocation +memory.grow+ delta past the entry-time
    # baseline). Returns +Kobako::Usage::EMPTY+ before any invocation;
    # populated on every outcome — including +TrapError+ — so the Host
    # App can read it after rescuing a trap to diagnose budget
    # consumption.
    def usage = @last_context.usage

    # Build a fresh Sandbox.
    #
    # +wasm_path+ is the absolute path to the Guest Binary; defaults to the
    # gem-bundled +data/kobako.wasm+. Every other keyword — the four caps
    # (+stdout_limit+, +stderr_limit+, +timeout+, +memory_limit+) and the
    # requested isolation profile (+profile+) — is forwarded verbatim to
    # +Kobako::SandboxOptions+, which owns the DEFAULT fallbacks and
    # normalisation. The constructed +SandboxOptions+ is exposed as
    # +#options+ and every option remains readable directly on Sandbox via
    # +Forwardable+ delegation. The runtime builds the requested profile —
    # +:hermetic+ (the default) denies the guest ambient time and entropy,
    # +:permissive+ leaves them live — and construction refuses a runtime
    # whose declared profile falls below the request, raising
    # +Kobako::SetupError+ before any invocation entry point runs.
    def initialize(wasm_path: nil, **)
      @wasm_path = wasm_path || Kobako::Runtime.default_path
      @options = SandboxOptions.new(**)
      @services = Kobako::Catalog::Services.new
      @snippets = Catalog::Snippets.new
      @extensions = Catalog::Extensions.new
      @runtime = build_runtime!
      reset_invocation_state!
    end

    # Bind +object+ as the Service reachable at +path+ — a Symbol or
    # String of one or more +::+-separated constant-form segments
    # (+"MyService::KV"+ or a top-level +"File"+). Returns +self+ for
    # chaining.
    #
    # Called with only a +path+, it declares a fillable Service:
    # +bind(path)+ reserves the path for +Kobako::Unresolved+, so the guest
    # sees the constant while the host defers the object it stands for. A
    # guest dispatch to an unfilled fillable surfaces as
    # +Kobako::ServiceError+ when left unrescued.
    #
    # Raises +ArgumentError+ when a segment is malformed, when +path+
    # collides with an existing binding (a name is a bound Service or a
    # grouping prefix, never both), or when called after the first
    # invocation has sealed Service registration.
    def bind(path, object = Unresolved)
      @services.bind(path, object)
      self
    end

    # Install one or more Extensions — each a guest idiom (+source+) paired
    # with an optional host +backend+, composed onto the Sandbox through
    # +#preload+ and +#bind+. An Extension is any object exposing
    # +name+ / +source+ / +backend+ / +depends_on+; +Kobako::Extension+ is
    # the bundled value type. Returns +self+.
    #
    # Raises +ArgumentError+ for a malformed Extension, a call after the
    # first invocation seals registration, or — at that first invocation —
    # an unmet +depends_on+.
    def install(*extensions)
      raise ArgumentError, "cannot install after first Sandbox invocation" if @services.sealed?

      extensions.each { |extension| @extensions.install(extension, snippets: @snippets, services: @services) }
      self
    end

    # Register a snippet on this Sandbox in one of two forms:
    #
    #   * +preload(code: source, name: Name)+ — +source+ is mruby source
    #     as a +String+ and +Name+ matches +/\A[A-Z]\w*\z/+. Compile
    #     failures surface as +Kobako::SandboxError+ on the first
    #     invocation's replay. The +name+
    #     becomes the snippet's +(snippet:Name)+ backtrace filename and
    #     is the dedupe key that rejects a duplicate +code:+ snippet.
    #   * +preload(binary: bytes)+ — +bytes+ is precompiled RITE
    #     bytecode as a +String+. The canonical name, when present,
    #     lives in the bytecode's embedded +debug_info+ and is resolved
    #     by the guest at load time; the host treats the bytes as
    #     opaque. Structural failures surface as +Kobako::BytecodeError+
    #     on the first invocation.
    #
    # Subsequent invocations (+#eval+ or +#run+) replay every registered
    # snippet — in insertion order — against the fresh +mrb_state+
    # before per-invocation source or entrypoint resolution.
    #
    # Returns +self+ to allow chaining.
    #
    # Raises +ArgumentError+ when neither form's keyword set is
    # supplied, when both forms are mixed (e.g., +code:+ and +binary:+
    # together, or +binary:+ paired with +name:+), when +code+ / +bytes+
    # is not a +String+, when +name+ does not match the constant
    # pattern, when +name+ duplicates an already-registered +code:+ form
    # snippet, or when called after the first invocation has sealed the
    # snippet table.
    def preload(code: nil, name: nil, binary: nil)
      raise ArgumentError, "cannot preload after first Sandbox invocation" if @services.sealed?

      @snippets.register(code: code, name: name, binary: binary)
      self
    end

    # Dispatch into a preloaded entrypoint constant. Delegates host
    # pre-flight and wire encoding to +Kobako::Transport::Run+ /
    # +Kobako::Transport::Run#encode+: a non-Symbol/String +target+ raises
    # +TypeError+, while a +target+ failing the constant pattern, a forged
    # +Kobako::Handle+ in +args+ / +kwargs+, or a non-Symbol +kwargs+ key
    # raise +ArgumentError+. The guest resolves +target+ as a top-level
    # constant, calls +#call+ on it with +args+ / +kwargs+, and returns the
    # deserialized result. The first invocation seals the Service registry
    # and snippet table. Runtime errors follow the same three-class
    # taxonomy as +#eval+.
    def run(target, *args, **kwargs, &block)
      run_envelope = Transport::Run.new(entrypoint: target, args: args, kwargs: kwargs)
      new_invocation.run(run_envelope, &block)
    end

    # Execute a guest mruby source string in a fresh +mrb_state+. +code+ is
    # the mruby source as a UTF-8 String. Returns the deserialized last
    # expression of the source.
    #
    # Source delivery uses the WASI stdin three-frame protocol
    # ({docs/wire-codec.md Invocation channels}[link:../../docs/wire-codec.md]):
    # Frame 1 carries the msgpack-encoded preamble (Service registry
    # snapshot), Frame 2 carries the user source UTF-8 bytes, and
    # Frame 3 carries the snippet table registered via +#preload+.
    # Each frame is prefixed by a 4-byte big-endian u32 length; Frame 3 is
    # mandatory-presence — an empty snippet table sends an empty msgpack
    # array, never an absent frame.
    #
    # The first invocation seals the Service registry and snippet table;
    # subsequent +#bind+ / +#preload+ calls raise +ArgumentError+.
    #
    # Raises +Kobako::TrapError+ on a Wasm trap or wire-violation fallback;
    # +Kobako::SandboxError+ when the guest ran to completion but failed
    # (including when +code+ is +nil+ or not a String, or when a preloaded
    # snippet's replay raises); +Kobako::ServiceError+ on an unrescued
    # Service capability failure.
    def eval(code, &block)
      raise SandboxError, "code must be a String, got #{code.class}" unless code.is_a?(String)

      new_invocation.eval(code, &block)
    end

    # Install a fresh, unrun +Context+ as the last invocation, so the
    # observable readers (+#stdout+ / +#stderr+ / +#usage+) return their
    # pre-invocation sentinels. Runs at construction and is called by
    # +Kobako::Pool+ at checkout so a pooled Sandbox hands over empty buffers.
    def reset_invocation_state!
      @last_context = Context.new(runtime: @runtime, services: @services, snippets: @snippets,
                                  extensions: @extensions)
    end

    private

    # Construct the +Runtime+ with the requested isolation profile and
    # refuse one whose declared posture falls below the request —
    # +SandboxOptions#enforce_floor!+ owns the ladder comparison, so a
    # runtime that cannot honor the request never runs guest code.
    def build_runtime!
      runtime = Kobako::Runtime.from_path(@wasm_path, @options.timeout, @options.memory_limit,
                                          @options.stdout_limit, @options.stderr_limit, @options.profile)
      @options.enforce_floor!(runtime.profile)
      runtime
    end

    # Seal the config on the first invocation and install a fresh
    # per-invocation +Context+ as the last invocation; returns it for the verb
    # to drive. The Context resolves its own callable Extension backends, so no
    # per-invocation state is written back onto the shared config here.
    def new_invocation
      begin_invocation!
      reset_invocation_state!
      @last_context
    end

    # Per-invocation prologue on the config tier: seals the Service / snippet /
    # Extension registries on the first call (idempotent — asserting Extension
    # dependencies then). Per-invocation provider resolution and observable
    # state live on the +Context+, not here.
    def begin_invocation!
      @services.seal!
      @snippets.seal!
      @extensions.seal!
    end
  end
end

File: crates/kobako/src/sandbox.rs
//! The Sandbox: one guest, its Services, and the invocation verbs.
//!
//! The Rust counterpart of `Kobako::Sandbox`: registrations and
//! preloads fill the Catalog until the first invocation seals it, then
//! `eval` / `run` execute on a fresh guest instance and return an
//! `Execution` — the per-invocation record carrying the decoded value
//! (or a guest failure's taxonomy `Error`), the output captures, and
//! usage. Per-invocation state — the capability-Handle table and the
//! observables — lives on that returned `Execution`, not on the reused
//! Sandbox, so nothing an invocation produces outlives it on `self`.

use std::path::Path;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use kobako_codec::codec::{Encode as _, Value};
use kobako_codec::payload::Arguments;
use kobako_runtime::envelope::Run;
use kobako_runtime::profile::Profile;
use kobako_runtime::runtime::{Entry, Frames, Runtime};
pub use kobako_runtime::snapshot::Usage;
use kobako_runtime::snapshot::{Completion, Snapshot};
use kobako_wasmtime::{Config, Driver};

use crate::catalog::Catalog;
use crate::dispatch::CatalogHandler;
use crate::error::{Error, GuestFailure};
use crate::execution::Execution;
use crate::extension::{install_object, unresolved, Extension, Extensions};
use crate::handles::{HandleTable, Handles};
use crate::outcome;
use crate::receiver::Receiver;
use crate::snippet;

/// Per-invocation path→object resolutions the dispatch handler answers ahead
/// of the sealed Catalog: the `ctx.bind` overrides followed by each
/// `PerInvocation` provider's fresh object.
type Resolved = Vec<(String, Arc<dyn Receiver>)>;

/// Per-Sandbox caps and posture, the counterpart of the Ruby
/// `SandboxOptions` value object. `None` means "no cap".
#[derive(Clone)]
pub struct Options {
    /// Wall-clock cap for one invocation.
    pub timeout: Option<Duration>,
    /// Guest linear-memory cap, in bytes.
    pub memory_limit: Option<usize>,
    /// Captured-stdout cap, in bytes.
    pub stdout_limit: Option<usize>,
    /// Captured-stderr cap, in bytes.
    pub stderr_limit: Option<usize>,
    /// Requested isolation floor; the driver declares its posture and
    /// construction fails below the floor.
    pub profile: Profile,
}

impl Default for Options {
    /// Secure by default: no caps, hermetic floor — the same default
    /// posture the Ruby frontend requests.
    fn default() -> Self {
        Options {
            timeout: None,
            memory_limit: None,
            stdout_limit: None,
            stderr_limit: None,
            profile: Profile::Hermetic,
        }
    }
}

/// The registration tables' seal-once lifecycle: open for setup, then
/// immutable from the first invocation on.
enum Registry {
    Open(Catalog),
    Sealed(Arc<Catalog>),
}

impl Registry {
    /// Mutate the open catalog, or refuse once sealed.
    fn open_mut(&mut self) -> Result<&mut Catalog, Error> {
        match self {
            Registry::Open(catalog) => Ok(catalog),
            Registry::Sealed(_) => Err(Error::Sealed(
                "registrations must happen before the first invocation",
            )),
        }
    }

    /// Seal on first use and hand out the shared table.
    fn seal(&mut self) -> Arc<Catalog> {
        if let Registry::Open(catalog) = self {
            let sealed = Arc::new(std::mem::take(catalog));
            *self = Registry::Sealed(sealed);
        }
        match self {
            Registry::Sealed(catalog) => catalog.clone(),
            Registry::Open(_) => unreachable!("seal above pinned the sealed state"),
        }
    }
}

/// A `run` argument: a `Value` passes by value, a host object
/// auto-wraps into a capability Handle the guest can call back into
/// (the counterpart of the Ruby `#run` auto-wrap; wrapping applies to
/// the top-level argument position).
pub enum RunArg {
    Value(Value),
    Object(Arc<dyn Receiver>),
}

impl From<Value> for RunArg {
    fn from(value: Value) -> Self {
        RunArg::Value(value)
    }
}

/// One guest sandbox: construction loads the Guest Binary and each
/// `eval` / `run` invokes it, returning that invocation's `Execution`.
/// The reused Sandbox holds only sealed config — no per-invocation
/// state — so it is the config tier and the `Execution` is the result.
pub struct Sandbox {
    driver: Driver,
    // Behind a `Mutex` so the first `eval` / `run` can seal it through
    // `&self`: setup verbs reach it with `get_mut` (they hold `&mut self`,
    // before the Sandbox is shared), the seal locks it. `Send + Sync`, so
    // an `Arc<Sandbox>` drives concurrent `&self` invocations.
    registry: Mutex<Registry>,
    extensions: Extensions,
}

impl Sandbox {
    /// Load a Guest Binary under the given caps. Fails with
    /// `Error::Setup` when the artifact is absent or unusable, or when
    /// the driver's declared posture falls below the requested floor.
    pub fn new(wasm_path: impl AsRef<Path>, options: Options) -> Result<Self, Error> {
        let config = Config {
            timeout: options.timeout,
            stdout_limit_bytes: options.stdout_limit,
            stderr_limit_bytes: options.stderr_limit,
            profile: options.profile,
        };
        let driver =
            Driver::new(wasm_path.as_ref(), options.memory_limit, config).map_err(Error::Setup)?;
        Ok(Sandbox {
            driver,
            registry: Mutex::new(Registry::Open(Catalog::default())),
            extensions: Extensions::default(),
        })
    }

    /// The open catalog for a setup mutation; refused once sealed. Setup
    /// runs on `&mut self`, before the Sandbox is shared, so it reaches the
    /// registry without locking.
    fn open_catalog(&mut self) -> Result<&mut Catalog, Error> {
        self.registry
            .get_mut()
            .expect("the registry mutex is never poisoned")
            .open_mut()
    }

    /// Bind a host object as the Service reachable at `path` — a
    /// constant path of one or more `::`-separated segments
    /// (`"MyService::KV"` or a top-level `"File"`). Refused once sealed.
    pub fn bind(&mut self, path: &str, object: Arc<dyn Receiver>) -> Result<(), Error> {
        self.open_catalog()?.bind(path, object);
        Ok(())
    }

    /// Declare a fillable Service at `path` with no object — the Rust spelling
    /// of the Ruby frontend's `bind(path)`. The path enters Frame 1 so the
    /// guest sees the constant, but is backed by the unresolved sentinel until
    /// an override fills it; a guest dispatch to an unfilled fillable fails
    /// closed as an undefined target (a guest `ServiceError`). Refused once
    /// sealed.
    pub fn bind_fillable(&mut self, path: &str) -> Result<(), Error> {
        self.open_catalog()?.bind(path, unresolved());
        Ok(())
    }

    /// Install an Extension — a guest idiom (`source`) paired with an
    /// optional host backend — composing it over `preload` and `bind`. A
    /// `Static` provider binds its object directly; a `PerInvocation`
    /// provider is resolved fresh at every invocation. Refused once the
    /// first invocation seals registration; an Extension whose `depends_on`
    /// names one that was not installed surfaces at that first invocation.
    pub fn install(&mut self, extension: Arc<dyn Extension>) -> Result<(), Error> {
        let catalog = self.open_catalog()?;
        catalog
            .snippets
            .register_source(extension.name(), extension.source())?;
        if let Some(backend) = extension.backend() {
            catalog.bind(&backend.path, install_object(&backend.provider));
        }
        self.extensions.record(extension);
        Ok(())
    }

    /// Register a source snippet for per-invocation replay under its
    /// canonical backtrace name. Refused once sealed, on a
    /// non-constant name, or on a duplicate name.
    pub fn preload(&mut self, name: &str, source: &str) -> Result<(), Error> {
        self.open_catalog()?.snippets.register_source(name, source)
    }

    /// Register precompiled RITE bytecode for per-invocation replay.
    /// The bytes stay opaque host-side; the guest validates them at
    /// first replay. Refused once sealed.
    pub fn preload_binary(&mut self, bytecode: impl Into<Vec<u8>>) -> Result<(), Error> {
        self.open_catalog()?
            .snippets
            .register_binary(bytecode.into());
        Ok(())
    }

    /// Run one mruby source on a fresh guest instance and return its
    /// `Execution` — the record of the run. `Ok` means the guest ran (its
    /// last expression, or a guest failure, rides `Execution::value`); the
    /// outer `Err` means it never started.
    pub fn eval(&self, source: &str) -> Result<Execution, Error> {
        let (catalog, handles) = self.begin_invocation()?;
        self.invoke(
            catalog,
            handles,
            Entry::Eval {
                source: source.as_bytes(),
            },
            Vec::new(),
        )
    }

    /// `eval` with a per-invocation override closure — the Rust spelling of
    /// the Ruby frontend's `#eval { |ctx| ctx.bind(...) }`. The closure runs
    /// before the guest drives, receiving the per-invocation `Context` whose
    /// `bind` fills a fillable or shadows any declared binding for this
    /// invocation only; overriding an undeclared path returns `Error::Argument`
    /// before the guest runs. An override takes priority over that path's
    /// per-invocation provider result and its static base, and touches
    /// host-side resolution only — Frame 1 stays fixed.
    pub fn eval_with<F>(&self, source: &str, overrides: F) -> Result<Execution, Error>
    where
        F: FnOnce(&mut Context<'_>) -> Result<(), Error>,
    {
        let (catalog, handles) = self.begin_invocation()?;
        let resolved = collect_overrides(&catalog, overrides)?;
        self.invoke(
            catalog,
            handles,
            Entry::Eval {
                source: source.as_bytes(),
            },
            resolved,
        )
    }

    /// Dispatch into a preloaded entrypoint with positional and keyword
    /// arguments; the guest resolves `target` as a top-level constant and
    /// invokes its `call`. A `RunArg::Object` argument auto-wraps into a
    /// capability Handle before the envelope encodes. Host pre-flight refuses a
    /// non-constant `target` before the invocation seals the tables, matching
    /// the Ruby frontend's ordering.
    pub fn run(
        &self,
        target: &str,
        args: Vec<RunArg>,
        kwargs: Vec<(String, RunArg)>,
    ) -> Result<Execution, Error> {
        self.drive_run(target, args, kwargs, |_| Ok(Vec::new()))
    }

    /// `run` with a per-invocation override closure — the Rust spelling of the
    /// Ruby frontend's `#run(target, ...) { |ctx| ctx.bind(...) }`, the `run`
    /// counterpart of `eval_with`. The closure runs before the guest drives and
    /// binds overrides under the same rules `eval_with` documents.
    pub fn run_with<F>(
        &self,
        target: &str,
        args: Vec<RunArg>,
        kwargs: Vec<(String, RunArg)>,
        overrides: F,
    ) -> Result<Execution, Error>
    where
        F: FnOnce(&mut Context<'_>) -> Result<(), Error>,
    {
        self.drive_run(target, args, kwargs, move |catalog| {
            collect_overrides(catalog, overrides)
        })
    }

    /// Shared `run` / `run_with` core: validate the target before sealing, seal,
    /// collect any overrides against the sealed catalog, auto-wrap the args into
    /// this run's Handle table, and drive the entrypoint envelope. `collect`
    /// yields the per-invocation overrides — empty for `run`, the closure's for
    /// `run_with`.
    fn drive_run<C>(
        &self,
        target: &str,
        args: Vec<RunArg>,
        kwargs: Vec<(String, RunArg)>,
        collect: C,
    ) -> Result<Execution, Error>
    where
        C: FnOnce(&Catalog) -> Result<Resolved, Error>,
    {
        if !snippet::constant_name(target) {
            return Err(Error::Argument(format!(
                "entrypoint must be a Ruby constant name (got {target:?})"
            )));
        }
        let (catalog, handles) = self.begin_invocation()?;
        let resolved = collect(&catalog)?;
        let args = args
            .into_iter()
            .map(|arg| wrap_run_arg(&handles, arg))
            .collect::<Result<_, _>>()?;
        let kwargs = kwargs
            .into_iter()
            .map(|(key, arg)| Ok((key, wrap_run_arg(&handles, arg)?)))
            .collect::<Result<_, Error>>()?;
        let payload = Arguments::new(args, kwargs)
            .encode()
            .map_err(|err| Error::Argument(format!("arguments are not wire-encodable: {err}")))?;
        let envelope = Run {
            entrypoint: target.to_string(),
            payload,
        }
        .encode();
        self.invoke(
            catalog,
            handles,
            Entry::Run {
                envelope: &envelope,
            },
            resolved,
        )
    }

    /// Shared invocation core behind `eval` / `run`: assemble the
    /// sealed catalog's frames and dispatch handler over this invocation's
    /// fresh Handle table, drive `entry` through the driver, and cook the
    /// snapshot into an `Execution` — one owner for the wiring so a handler
    /// or frame change cannot drift between verbs. `&self` because no
    /// per-invocation state is written back: the `handles` table and the
    /// snapshot's observables ride into the returned `Execution`.
    /// `resolved` starts with the per-invocation `ctx.bind` overrides so the
    /// dispatch handler answers them before this run's provider results.
    fn invoke(
        &self,
        catalog: Arc<Catalog>,
        handles: Arc<Mutex<HandleTable>>,
        entry: Entry<'_>,
        mut resolved: Resolved,
    ) -> Result<Execution, Error> {
        resolved.extend(self.extensions.resolve());
        let preamble = catalog.preamble();
        let snippets = catalog.snippets.frame();
        let handler = Arc::new(CatalogHandler::new(catalog, handles.clone(), resolved));
        let snapshot = self.driver.invoke(
            entry,
            Frames {
                preamble: &preamble,
                snippets: &snippets,
            },
            Some(handler),
        )?;
        Ok(build_execution(snapshot, handles))
    }

    /// Per-invocation prologue on `&self`: seal the registration tables and
    /// assert Extension dependencies on the first invocation, then hand back
    /// the sealed catalog and a fresh Handle table this invocation owns. The
    /// seal locks the registry, so concurrent first invocations serialize on
    /// it and all observe the same sealed catalog. An unmet dependency raises
    /// before the guest runs.
    fn begin_invocation(&self) -> Result<(Arc<Catalog>, Arc<Mutex<HandleTable>>), Error> {
        let catalog = self
            .registry
            .lock()
            .expect("the registry mutex is never poisoned")
            .seal();
        self.extensions.assert_dependencies()?;
        Ok((catalog, Arc::new(Mutex::new(HandleTable::default()))))
    }
}

/// Cook a raw `Snapshot` into the invocation's `Execution`: captures and
/// usage carry over verbatim, and the completion becomes the guest-level
/// `outcome` — a decoded value (whose Handles must all be live), or the
/// taxonomy `Error` a trap or guest failure attributes to. The `handles`
/// table rides along so the result's Handles resolve on the Execution.
fn build_execution(snapshot: Snapshot, handles: Arc<Mutex<HandleTable>>) -> Execution {
    let outcome = match snapshot.completion {
        Completion::Outcome(bytes) => outcome::decode(&bytes).and_then(|value| {
            require_live_handles(&handles, &value)?;
            Ok(value)
        }),
        Completion::Trap(trap) => Err(trap.into()),
    };
    Execution::new(
        outcome,
        handles,
        snapshot.stdout,
        snapshot.stderr,
        snapshot.usage,
    )
}

/// Encode one `run` argument, auto-wrapping a host object into the
/// invocation's Handle table. Exhaustion surfaces pre-call with the Ruby
/// counterpart's attribution — an outer `Err`, since the guest never ran.
fn wrap_run_arg(handles: &Mutex<HandleTable>, arg: RunArg) -> Result<Value, Error> {
    match arg {
        RunArg::Value(value) => Ok(value),
        RunArg::Object(object) => handles
            .lock()
            .expect("the Handle table mutex is never poisoned")
            .alloc(object)
            .map(Value::Handle)
            .map_err(|message| {
                Error::Sandbox(GuestFailure {
                    class: "Kobako::HandleExhaustedError".into(),
                    message,
                    backtrace: Vec::new(),
                    details: None,
                })
            }),
    }
}

/// Every Handle a guest legitimately returns resolves to a live object
/// (it cannot fabricate one); an unknown id in the result signals a
/// corrupted runtime and fails like a malformed value.
fn require_live_handles(handles: &Mutex<HandleTable>, value: &Value) -> Result<(), Error> {
    match value {
        Value::Handle(id) => {
            if Handles::new(handles).resolve(value).is_some() {
                Ok(())
            } else {
                Err(Error::Sandbox(GuestFailure {
                    class: "Kobako::SandboxError".into(),
                    message: format!("unknown Handle id: {id}"),
                    backtrace: Vec::new(),
                    details: None,
                }))
            }
        }
        Value::Array(items) => items
            .iter()
            .try_for_each(|v| require_live_handles(handles, v)),
        Value::Map(pairs) => pairs.iter().try_for_each(|(key, val)| {
            require_live_handles(handles, key)?;
            require_live_handles(handles, val)
        }),
        _ => Ok(()),
    }
}

/// Run an override closure against a fresh `Context` over `catalog` and hand
/// back the `ctx.bind` overrides it collected — the step `eval_with` and
/// `run_with` share so the two verbs bind overrides identically.
fn collect_overrides<F>(catalog: &Catalog, overrides: F) -> Result<Resolved, Error>
where
    F: FnOnce(&mut Context<'_>) -> Result<(), Error>,
{
    let mut ctx = Context {
        catalog,
        overrides: Vec::new(),
    };
    overrides(&mut ctx)?;
    Ok(ctx.overrides)
}

/// The per-invocation Context handed to an `eval_with` / `run_with` override closure — the
/// Rust peer of the Ruby frontend's `Kobako::Context`. Here it carries only the
/// `ctx.bind` overrides; the run's Handle table and observables ride into its
/// `Execution` instead. Each override takes priority over that path's
/// per-invocation provider result and its static base for that one invocation,
/// and is discarded when it ends. The borrow keeps it from outliving the
/// closure, which is what spends it.
pub struct Context<'a> {
    catalog: &'a Catalog,
    overrides: Resolved,
}

impl Context<'_> {
    /// Override the object bound at an already-declared `path` for this
    /// invocation — filling a fillable or shadowing a static / per-invocation
    /// binding. Overriding a path that was never declared returns
    /// `Error::Argument`, so an override can never grow the sealed key set.
    /// A second override of the same `path` wins over the first, matching the
    /// Ruby frontend's last-wins semantics — an override is the caller's final
    /// word on that path for the run.
    pub fn bind(&mut self, path: &str, object: Arc<dyn Receiver>) -> Result<(), Error> {
        if self.catalog.lookup(path).is_none() {
            return Err(Error::Argument(format!(
                "cannot override undeclared path {path:?}"
            )));
        }
        self.overrides.retain(|(bound, _)| bound != path);
        self.overrides.push((path.to_string(), object));
        Ok(())
    }
}

/// `Arc<Sandbox>` drives concurrent `&self` invocations, which requires
/// `Sandbox: Send + Sync`. Assert it at compile time so a future field that
/// breaks the property fails the build here, not at a distant call site.
const _: fn() = || {
    fn assert_send_sync<T: Send + Sync>() {}
    assert_send_sync::<Sandbox>();
};

#[cfg(test)]
mod tests {
    use super::*;

    // The seal-once lifecycle is pure state and testable without a
    // driver; the invocation path itself is pinned end-to-end by the
    // parity harness against the real guest binary.
    #[test]
    fn registry_seals_once_and_refuses_late_mutation() {
        let mut registry = Registry::Open(Catalog::default());
        assert!(registry.open_mut().is_ok());
        let first = registry.seal();
        let second = registry.seal();
        assert!(Arc::ptr_eq(&first, &second));
        assert!(matches!(registry.open_mut(), Err(Error::Sealed(_))));
    }
}
